articles/intraprocess_communication.md

```diff
@@ -0,0 +1,494 @@
+
+6. `Publisher::publish(std::unique_ptr<MessageT> msg)` calls `Publisher::do_inter_process_publish(const MessageT & inter_process_msg)`, where `MessageT inter_process_msg = *shared_msg`.
+
+The difference from the previous case is that here a `std::shared_ptr<const MessageT>` is being "added" to the buffers.
+Note that this `std::shared_ptr` has been just created from a `std::unique_ptr` and it is only used by the `IntraProcessManager` and by the RMW, while the user application has no access to it.
+
+ - `BufferT = unique_ptr<MessageT>` The buffer receives a copy of `MessageT` and has ownership on it.
+ - `BufferT = shared_ptr<const MessageT>` Every buffer receives a shared pointer of the same `MessageT`, so no copies are required.
+ - `BufferT = MessageT` A copy of the message is added to every buffer.
+
+The difference with publishing a unique_ptr is that here it is not possible to move the ownership of the message to one of the `Subscription`, potentially saving a copy.
+
+![Sequence UML diagram](../img/intraprocess_communication/intra_inter_process.png)
+
+### QoS features
+
+The proposed implementation can handle all the different QoS.
+
+ - If the history is set to `keep_last`, then the depth of the history corresponds to the size of the ring buffer.
+ On the other hand, if the history is set to `keep_all`, the buffer becomes a standard FIFO queue with an unbounded size.
+ - The reliability is only checked by the `IntraProcessManager` in order to understand if a `Publisher` and a `Subscription` are compatible.
+ The use of buffers ensures that all the messages are delivered without the need to resend them.
+ Thus, both options, `reliable` and `best-effort`, are satisfied.
+ - The durability QoS is used to understand if a `Publisher` and a `Subscription` are compatible.
+ How this QoS is handled is described in details in the following paragraph.
+
+#### Handling Transient Local
+
+If the `Publisher` durability is set to `transient_local` an additional buffer on the `Publisher` side is used to store the sent intra-process messages.
+
+Late-joiner `Subscription`s will have to extract messages from this buffer once they are added to the `IntraProcessManager`.
+In this case the `IntraProcessManager` has to check if the recently created `Subscription` is a late-joiner, and, if it is, it has to retrieve messages from the `Transient Local` `Publisher`s.
+
+1. Call `IntraProcessManager::find_matching_publishers(SubscriptionInfo sub_info)` that returns a list of stored `PublisherInfo` that have a QoS compatible for sending messages to this new `Subscription`.
+These will be all `Transient Local` `Publisher`s, so they have a ring buffer.
+2. Copy messages from all the ring buffers found into the ring buffer of the new `Subscription`.
+**TODO:** are there any constraints on the order in which old messages have to be retrieved? (i.e. 1 publisher at the time; all the firsts of each publisher, then all the seconds ...).
```

Comment:
I don't know if there is an order because I think it may be asynchronous when DDS does this, it "requests" the back filled messages and then they arrive later. See this conversation starting here:

https://github.com/ros2/geometry2/pull/160#pullrequestreview-285066035